sys/include/barriers.h

```diff
@@ -0,0 +1,298 @@
+/*
+ * Copyright (C) 2019 Otto-von-Guericke-Universität Magdeburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    sys_barriers Memory Barriers
+ * @ingroup     sys
+ * @brief       Memory Barriers
+ *
+ * # FAQ
+ * ## What are Memory Barriers
+ * Memory barriers suppress specific kinds optimization performed by the
+ * compiler or (in case of out-of-order CPUs) CPUs, depending on the barrier.
+ * This will most likely result in slower program execution and, thus, is in
+ * general unwanted. However, under specific circumstances, optimization can
+ * prevent correct execution.
+ *
+ * ## Why can't I use `volatile` instead
+ * Memory accesses to `volatile` can be viewed as an alternative to
+ * @ref compiler_barrier, but only in regard to other accesses to `volatile`
+ * variables. (Some compilers do not reorder memory accesses to non-`volatile`s
+ * across accesses to `volatile`s as well, but not all compilers do so.)
+ * Resorting to mark all variables as `volatile` to enforce the correct order
+ * will prevent any reordering of any access to those variables, and thus have
+ * a bigger impact on performance than using barriers. Additionally those are
+ * only compiler barriers that cannot replace @ref memory_barrier.
+ *
+ * ## When do I need to suppress optimization
+ * Generally speaking, when memory is accesses outside of the current program
+ * flow. This will happen if another thread, an interrupt service handler, or
+ * hardware is accessing the same memory.
+ *
+ * ### Hardware mapped to memory
+ * #### TL;DR
+ * Declare memory directly accessed hardware as `volatile`. In addition the use
+ * of memory barrier can be necessary, unless reordering of any memory access
+ * to non-`volatile` memory is acceptable. Examples of access requiring
+ * barriers additionally include the access to timer registers or interrupt
+ * control registers.
+ *
+ * #### In detail
+ * Accesses to memory that is directly interacting with hardware, such as timer
+ * registers, GPIO registers, or interrupt control registers get easily
+ * optimized out by the compiler. Lets assume an 16-bit timer register is
+ * located at address `0x42` and we want to use that to measure the time a call
+ * to `void do_computing(void)` takes:
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
+ * // BEWARE: This example deliberately contains **BUGS**
+ * #define TIMER    (*((uint16_t *)0x42))
+ *
+ * uint16_t measure_time(void) {
+ *     uint16_t pre = TIMER;
+ *     do_computing();
+ *     uint16_t post = TIMER;
+ *     return pre - post;
+ * }
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * In the above example from the compilers perspective the value in `TIMER`
```

Comment:
Have you seen this behaviour in practice?

Reply:
I did. This is also the example presented here https://blog.regehr.org/archives/28, or similar here https://hackaday.com/2015/08/18/embed-with-elliot-the-volatile-keyword/